backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/questionsubmission/QuestionSubmissionService.java

```diff
@@ -0,0 +1,278 @@
+
+        Review review = new Review(user, questionSubmission, reviewDto);
+
+        updateQuestionStatus(review.getStatus(), questionSubmission.getQuestion().getId());
+
+        reviewRepository.save(review);
+        return new ReviewDto(review);
+    }
+
+    @Retryable(value = {SQLException.class}, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public QuestionSubmissionDto updateQuestionSubmission(Integer questionSubmissionId, QuestionSubmissionDto questionSubmissionDto) {
+        QuestionSubmission questionSubmission = getQuestionSubmission(questionSubmissionId);
+
+        User user = getStudent(questionSubmissionDto.getUserId());
+
+        if(user.isStudent() && questionSubmission.getQuestion().getStatus() != Question.Status.IN_REVISION) {
+            throw new TutorException(CANNOT_EDIT_REVIEWED_QUESTION);
+        }
+
+        this.questionService.updateQuestion(questionSubmissionDto.getQuestion().getId(), questionSubmissionDto.getQuestion());
+        return new QuestionSubmissionDto(questionSubmission);
+    }
+
+    @Retryable(
+            value = { SQLException.class },
+            backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.READ_COMMITTED)
+    public void updateQuestionSubmissionTopics(Integer questionSubmissionId, TopicDto[] topics) {
+        Integer questionId = questionSubmissionRepository.findQuestionIdByQuestionSubmissionId(questionSubmissionId);
+
+        questionService.updateQuestionTopics(questionId, topics);
+    }
+
+    @Retryable(
+            value = { SQLException.class },
+            backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.READ_COMMITTED)
+    public void removeSubmittedQuestion(Integer questionSubmissionId) {
+        QuestionSubmission questionSubmission = questionSubmissionRepository.findById(questionSubmissionId).orElseThrow(() -> new TutorException(QUESTION_SUBMISSION_NOT_FOUND, questionSubmissionId));
+        Question question = questionRepository.findById(questionSubmission.getQuestion().getId()).orElseThrow(() -> new TutorException(QUESTION_NOT_FOUND, questionSubmission.getQuestion().getId()));
+
+        if(!questionSubmission.getReviews().isEmpty() || !question.getStatus().equals(Question.Status.IN_REVISION)) {
+            throw new TutorException(CANNOT_DELETE_REVIEWED_QUESTION);
+        } else {
+            removeQuestionSubmission(questionSubmission.getId());
+            question.remove();
+            questionRepository.delete(question);
+        }
+    }
+
+    @Retryable(value = {SQLException.class}, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public void removeQuestionSubmission(Integer questionSubmissionId) {
+        QuestionSubmission questionSubmission = getQuestionSubmission(questionSubmissionId);
+
+        removeReviews(questionSubmission);
+
+        questionSubmission.remove();
```

Comment:
These two removes look redundant, the remove of the question submission should remove the reviews

Reply:
but since remove does not actually delete it from the repository, I refactored the code that removes and then deletes the review from the repository, is it bad practice? or should reviews be removed in question submission's remove() and then delete all Reviews from repository after?

Reply:
now thinking about it, you can never eliminate a submission that has reviews, so maybe deleting reviews should only be done in question submission's reviews as stated